### PR TITLE
Fix 6481 - unbounded growth in BackupList from failure retries.

### PR DESCRIPTION
**Pull request by james-munson**
[BUG] Longhorn Instance Manager Memory leak #6481 

Three changes:  
 - Put in a log to catch runaway growth.
 - Ensure only one list entry per customer (backup ID).
 - Do the prune logic for "error" as well as "completed" status.

### CODE REVIEW

File: pkg/sync/rpc/list.go [hunk 1]
```diff
@@ -42,13 +52,24 @@ func (b *BackupList) BackupAdd(backupID string, backup *replica.BackupStatus) er
...
 	b.Lock()
 	b.infos = append(b.infos, &BackupInfo{
 		backupID:     backupID,
 		backupStatus: backup,
 	})
 	b.Unlock()
 
+	if backupListHighWaterMark < len(b.infos) {
+		backupListHighWaterMark = len(b.infos)
+		if 0 == backupListHighWaterMark%BackupListHighWaterMarkQuantum {
```
Comment (ejweber) [inline pkg/sync/rpc/list.go:68]: I'm not sure I understand this. This only returns `true` if `backupListHighWaterMark` is a multiple of 10, right? Are we trying to limit how often we log (only once every ten increases)?

Comment (james-munson) [reply pkg/sync/rpc/list.go:68]: Yep.  If we do have 1000 backups (possible) we don't really want to log 1000 times, but we don want to know if we've got a runaway list.  So this is just a simple-minded throttle.

Comment (ejweber) [reply pkg/sync/rpc/list.go:68]: Got it. Is there any known way to have 1000 backups after these changes? Shouldn't we delete any more than 10 error and 5 completed in `refresh` on 73?

(Not saying I have a problem with the throttle. Just making sure I understand our expectations.)

File: pkg/sync/rpc/list.go [hunk 2]
```diff
@@ -90,32 +111,34 @@ func (*BackupList) remove(b []*BackupInfo, index int) ([]*BackupInfo, error) {
...
 // Refresh deletes all the old completed backups from the front. Old backups are the completed backups
-// that are created before MaxBackupSize completed backups
+// that are created before the number we choose to retain.
 func (b *BackupList) refresh() error {
 	b.Lock()
 	defer b.Unlock()
 
-	var index, completed int
+	for state, limit := range retainBackupStateCounts {
+		var index, completed int
```
Comment (PhanLe1010) [inline pkg/sync/rpc/list.go:120]: The variable `completed` is a bit confusing. It used to count both `error` and `completed` backups. Should we change it to `count` instead?

**Review by PhanLe1010 — APPROVED** [review]
LGTM except for a small NIT

**Comment by PhanLe1010** [comment]
Could we run a e2e backup related tests for this one before merging? It is even better if run the full e2e test

File: pkg/sync/rpc/list.go (hunk 1 above)
Comment (james-munson) [reply pkg/sync/rpc/list.go:68]: It is certainly possible to have 1000 backups in progress - not yet either complete or error.  The refresh function only prunes those backups that have reached a conclusion.

**Review by ejweber — APPROVED** [review]
LGTM, question was answered.

I agree with @PhanLe1010 that we should link in a successful run of the end-to-end backup tests before merging.

**Comment by derekbit** [comment]
@james-munson 
Shall we remove the `backup` from the list if failed to`DoBackupCreate`?
https://github.com/longhorn/longhorn-engine/blob/master/pkg/sync/rpc/server.go#L721

**Comment by PhanLe1010** [comment]
Looks like this is closed in favor of https://github.com/longhorn/longhorn-engine/pull/927

**Pull request closed without merging**